### PR TITLE
FIR: check DELEGATE_SPECIAL_FUNCTION_RETURN_TYPE_MISMATCH

### PR DESCRIPTION
**Pull request by tgeng**

### CODE REVIEW

**Comment by jsjeon** [comment]
A [difference](https://punzki.github.io/Kotlin-DiagnosticTestsDiff/2021-04-13/viewDiff.html?expected=tests%2FdelegatedProperty%2FprovideDelegate%2FunsupportedOperatorProvideDelegate.kt&actual=tests%2FdelegatedProperty%2FprovideDelegate%2FunsupportedOperatorProvideDelegate.fir.kt) indicates that, with old inference, `DELEGATE_SPECIAL_FUNCTION_RETURN_TYPE_MISMATCH` is reported, whereas, with new inference, `DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE` is reported, which is the case for now (since FIR uses NI). I wonder we need to double-check if we can drop this diagnostic?

File: compiler/testData/diagnostics/tests/delegatedProperty/typeMismatchForGetReturnType.fir.kt [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 import kotlin.reflect.KProperty
 
-val c: Int by Delegate()
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/delegatedProperty/typeMismatchForGetReturnType.fir.kt:6]: Rather I'm curious why we missed `DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE`: https://github.com/JetBrains/kotlin/blob/ad9b962536e0d0c148baf884111c2c97c2bc8f2c/compiler/testData/diagnostics/tests/delegatedProperty/typeMismatchForGetReturnType.kt#L6

File: compiler/testData/diagnostics/tests/delegatedProperty/inference/differentDelegatedExpressions.fir.kt [hunk 2]
```diff
@@ -12,8 +12,8 @@ class A(outer: Outer) {
 
     var b: String by  foo(getMyProperty())
     var r: String by  foo(outer.getContainer().getMyProperty())
-    var e: String by  <!DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE!>+ foo(getMyProperty())<!>
-    var f: String by  <!DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE!>foo(getMyProperty()) - 1<!>
+    var e: String by  <!DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE, DELEGATE_SPECIAL_FUNCTION_RETURN_TYPE_MISMATCH!>+ foo(getMyProperty())<!>
+    var f: String by  <!DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE, DELEGATE_SPECIAL_FUNCTION_RETURN_TYPE_MISMATCH!>foo(getMyProperty()) - 1<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/delegatedProperty/inference/differentDelegatedExpressions.fir.kt:16]: Hm, actually, neither is reported?

https://github.com/JetBrains/kotlin/blob/ad9b962536e0d0c148baf884111c2c97c2bc8f2c/compiler/testData/diagnostics/tests/delegatedProperty/inference/differentDelegatedExpressions.kt#L15-L16

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt [hunk 3]
```diff
@@ -47,8 +49,13 @@ object FirDelegatedPropertyChecker : FirPropertyChecker() {
             override fun visitElement(element: FirElement) = element.acceptChildren(this)
 
             override fun visitFunctionCall(functionCall: FirFunctionCall) {
-                val errorNamedReference = functionCall.calleeReference as? FirErrorNamedReference ?: return
-                if (errorNamedReference.source?.kind != FirFakeSourceElementKind.DelegatedPropertyAccessor) return
+                val hasReferenceError = checkFunctionReferenceErrors(functionCall)
+                if (isGet && !hasReferenceError) checkReturnType(functionCall)
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt:53]: I expected `DELEGATE_SPECIAL_FUNCTION_NONE_APPLICABLE` case will make `hasReferenceError` `true`, but how can both diagnostics be reported?

Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt:53]: One is reported on setValue and on is reported on getValue.

Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt:53]: Aha!

File: compiler/testData/diagnostics/tests/delegatedProperty/typeMismatchForGetReturnType.fir.kt (hunk 1 above)
Comment (tgeng) [reply compiler/testData/diagnostics/tests/delegatedProperty/typeMismatchForGetReturnType.fir.kt:6]: I think FE1.0 adds a return type constraint to the inference system and FIR is not doing that. IMO it's a bug that FE1.0 is considering return type when resolving method calls. But we can ask JetBrains for confirmation.

File: compiler/testData/diagnostics/tests/delegatedProperty/inference/differentDelegatedExpressions.fir.kt (hunk 2 above)
Comment (tgeng) [reply compiler/testData/diagnostics/tests/delegatedProperty/inference/differentDelegatedExpressions.fir.kt:16]: Aha! Thanks! So it looks FE1.0 suppress reporting in this case when the inference for type parameters failed. Let me try mimic this behavior.

Comment (tgeng) [reply compiler/testData/diagnostics/tests/delegatedProperty/inference/differentDelegatedExpressions.fir.kt:16]: So we need to report NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER and suppress delegate issues in that case. But I just saw in https://docs.google.com/document/d/1f_uhx9LUqFD0Kl6V70rw03GFkBH3nVoGoA5RjcekNrM/edit#heading=h.j1zcgdakihi4 that says JetBrains folks will take care of it. Let me add a TODO here then.

**Review by jsjeon — APPROVED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt [hunk 4]
```diff
@@ -47,8 +52,13 @@ object FirDelegatedPropertyChecker : FirPropertyChecker() {
...
+            }
+
+            private fun checkFunctionReferenceErrors(functionCall: FirFunctionCall): Boolean {
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt:59]: Can you rename this to `hasFunctionReferenceErrors`?

**Review by punzki — APPROVED** [review]

**Comment by mglukhikh** [comment]
Merged manually 👍

**Pull request closed without merging**